src/main.h

```diff
@@ -399,8 +399,8 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
 bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex, CCoinsViewCache& coins, bool fJustCheck = false);
 
 /** Context-independent validity checks */
-bool CheckBlockHeader(const CBlockHeader& block, CValidationState& state, bool fCheckPOW = true);
-bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW = true, bool fCheckMerkleRoot = true);
+bool CheckBlockHeader(const CBlockHeader& block, CValidationState& state, bool fCheckPOW = true, bool protocol_cleanup = false);
+bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW = true, bool fCheckMerkleRoot = true, bool protocol_cleanup = false);
```

Comment:
Is it wise to allow a default value for the `protocol_cleanup` flag? It may be safer to have a compiler error when this is implicitly called as the consequences for an unhandled default using call could be rather severe after activation.

Reply:
True. I'll reorder those parameters and make it required.

Reply:
Done.

Reply:
Oops, not done. Fixing some of the calls reveals my mistake in organizing it this way: the checks affected by `protocol_cleanup` are now no longer context-free, and `CheckBlock()` and `CheckBlockHeader()` are called in places where there isn't chain history necessary to establish whether the protocol cleanup fork has activated.

This is a little tricky. What probably has to be done with the current state of things is to pull these checks into `ContextualCheckBlock` and `ContextualCheckBlockHeader` since that's when the block index context is available. In principle this is probably okay since we shouldn't be validating blocks that we don't have context for, thanks to headers-first synchronization, but I'm not 100% sure. Does this open up DoS vectors by allowing bad blocks to be relayed without header context? I don't know.

The other alternative is a soft-fork that I've been thinking about for some time -- require that `block.vtx[0].nLockTime == pindexPrev->GetMedianTimePast()`. This would make all contextual checks based on MTP become non-contextual -- the checks are made indirectly against the block coinbase's lock-time, which inself is required to be the median time past.

This would be rather straight forward to implement and deploy, and makes this PR considerably simpler.

Reply:
That's a cool idea! I'm not entirely sure I understand how it makes the MTP checks non-contextual, but will think on it.

Reply:
Anything that compares against median-time-past is rewritten to compare against the block's coinbase transaction's nLockTime. "Context" in this case means knowing the prior blocks or at least the block headers. That's necessary to check that the coinbase's nLockTime is, in fact, the median time past. But then all the other median time past checks can be done locally with no context other than the block itself.

Reply:
Aha! That's very useful indeed! Cool stuff.

Reply:
The code can be modified to do this even before deployment of the soft-fork, so I have edited it to use the Coinbase lock-time. I believe this issue is now resolved.